File: Q14 four sum.py
class Solution:
    def fourSum(self, nums, target):
        nums.sort()
        ans = []
        n = len(nums)

        for a in range(n):
            for b in range(a + 1, n):
                c, d = b + 1, n - 1
                while c < d:
                    total = nums[a] + nums[b] + nums[c] + nums[d]
                    if total < target:
                        c += 1
                    elif total > target:
                        d -= 1
                    else:
                        quad = [nums[a], nums[b], nums[c], nums[d]]
                        if quad not in ans:
                            ans.append(quad)
                        c += 1
                        d -= 1
        return ans

File: Q14 four sum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        int n = nums.length;
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);

        for (int a = 0; a < n; a++) {
            for (int b = a + 1; b < n; b++) {
                int c = b + 1;
                int d = n - 1;
                while (c < d) {
                    long sum = (long) nums[a] + nums[b] + nums[c] + nums[d];
                    if (sum < target) {
                        c++;
                    } else if (sum > target) {
                        d--;
                    } else {
                        List<Integer> v = Arrays.asList(nums[a], nums[b], nums[c], nums[d]);
                        if (!ans.contains(v)) {
                            ans.add(v);
                        }
                        c++;
                        d--;
                    }
                }
            }
        }
        return ans;
    }
}
